### PR TITLE
Fix scaled down deployments cannot identify old replica sets

### PR DESCRIPTION
**Pull request by shiywang**
Fixes https://github.com/kubernetes/kubernetes/issues/42570

### CODE REVIEW

File: pkg/controller/deployment/sync.go [hunk 1]
```diff
@@ -111,11 +111,11 @@ func (dc *DeploymentController) checkPausedConditions(d *extensions.Deployment)
 // This may lead to stale reads of replica sets, thus incorrect deployment status.
 func (dc *DeploymentController) getAllReplicaSetsAndSyncRevision(d *extensions.Deployment, rsList []*extensions.ReplicaSet, podMap map[types.UID]*v1.PodList, createIfNotExisted bool) (*extensions.ReplicaSet, []*extensions.ReplicaSet, error) {
 	// List the deployment's RSes & Pods and apply pod-template-hash info to deployment's adopted RSes/Pods
-	rsList, podList, err := dc.rsAndPodsWithHashKeySynced(d, rsList, podMap)
+	rsList, _, err := dc.rsAndPodsWithHashKeySynced(d, rsList, podMap)
```
Comment (kargakis) [inline pkg/controller/deployment/sync.go]: If we don't need to return the podList anymore, then we should remove it from the return arguments altogether.

File: pkg/controller/deployment/util/deployment_util.go [hunk 2]
```diff
@@ -807,22 +794,17 @@ func FindOldReplicaSets(deployment *extensions.Deployment, rsList []*extensions.
...
-			}
+	for _, rs := range rsList {
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		// Filter out new replica set
+		if newRS != nil && rs.UID == newRS.UID {
+			continue
+		}
+		allOldRSs[rs.ObjectMeta.Name] = rs
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: We don't need the maps but we can directly append to the slices, right?

Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: We don't need the maps and we can directly append to the slices, right?

File: pkg/controller/deployment/util/deployment_util.go [hunk 3]
```diff
@@ -807,22 +794,17 @@ func FindOldReplicaSets(deployment *extensions.Deployment, rsList []*extensions.
...
-			if newRS != nil && rs.UID == newRS.UID {
-				continue
-			}
-			// TODO: If there are no pods for a deployment, we will never return old replica sets....!
-			allOldRSs[rs.ObjectMeta.Name] = rs
-			if rsLabelsSelector.Matches(podLabelsSelector) {
-				oldRSs[rs.ObjectMeta.Name] = rs
-			}
+	for _, rs := range rsList {
+		if err != nil {
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: Redudant error check

File: pkg/controller/deployment/util/deployment_util.go [hunk 4]
```diff
@@ -807,22 +794,17 @@ func FindOldReplicaSets(deployment *extensions.Deployment, rsList []*extensions.
 	if err != nil {
 		return requiredRSs, allRSs, err
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: return explicit nils for the slices

**Comment by shiywang** [comment]
updated, ptal, thanks

**Comment by janetkuo** [comment]
@k8s-bot ok to test

**Comment by shiywang** [comment]
@kargakis @janetkuo  ptal

File: pkg/controller/deployment/util/deployment_util_test.go [hunk 5]
```diff
@@ -381,7 +381,7 @@ func TestGetOldRCs(t *testing.T) {
 		fakeClient = addGetRSReactor(fakeClient, test.objs[1])
 		fakeClient = addUpdatePodsReactor(fakeClient)
 		fakeClient = addUpdateRSReactor(fakeClient)
-		rss, _, err := GetOldReplicaSets(&newDeployment, fakeClient)
+		_, rss, err := GetOldReplicaSets(&newDeployment, fakeClient)
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util_test.go]: The pods used in this test are cruft now, can you remove them?

File: pkg/controller/deployment/util/deployment_util.go [hunk 6]
```diff
@@ -796,42 +783,22 @@ func FindNewReplicaSet(deployment *extensions.Deployment, rsList []*extensions.R
 
 // FindOldReplicaSets returns the old replica sets targeted by the given Deployment, with the given PodList and slice of RSes.
 // Note that the first set of old replica sets doesn't include the ones with no pods, and the second set of old replica sets include all old replica sets.
-func FindOldReplicaSets(deployment *extensions.Deployment, rsList []*extensions.ReplicaSet, podList *v1.PodList) ([]*extensions.ReplicaSet, []*extensions.ReplicaSet, error) {
-	// Find all pods whose labels match deployment.Spec.Selector, and corresponding replica sets for pods in podList.
-	// All pods and replica sets are labeled with pod-template-hash to prevent overlapping
-	oldRSs := map[string]*extensions.ReplicaSet{}
-	allOldRSs := map[string]*extensions.ReplicaSet{}
+func FindOldReplicaSets(deployment *extensions.Deployment, rsList []*extensions.ReplicaSet) ([]*extensions.ReplicaSet, []*extensions.ReplicaSet, error) {
 	requiredRSs := []*extensions.ReplicaSet{}
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: Change these to:
```go
var requiredRSs []*extensions.ReplicaSet
var allRSs []*extensions.ReplicaSet
```

File: pkg/controller/deployment/util/deployment_util_test.go [hunk 7]
```diff
@@ -585,52 +582,30 @@ func TestFindOldReplicaSets(t *testing.T) {
...
 			test:       "Get empty old ReplicaSets",
 			deployment: deployment,
 			rsList:     []*extensions.ReplicaSet{&newRS},
-			podList: &v1.PodList{
-				Items: []v1.Pod{
-					newPod,
-				},
-			},
-			expected: []*extensions.ReplicaSet{},
+			expected:   []*extensions.ReplicaSet{},
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util_test.go]: This should change to nil once you update the slices to be initialized with `var`

**Comment by shiywang** [comment]
@kargakis updated, ptal, thanks

File: pkg/controller/deployment/util/deployment_util_test.go [hunk 8]
```diff
@@ -301,23 +301,19 @@ func TestGetOldRCs(t *testing.T) {
 	newDeployment := generateDeployment("nginx")
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util_test.go]: Rename this test to `TestGetOldRSs`

File: pkg/controller/deployment/util/deployment_util_test.go [hunk 9]
```diff
@@ -585,52 +559,30 @@ func TestFindOldReplicaSets(t *testing.T) {
...
-				},
-			},
-			expected: []*extensions.ReplicaSet{},
+			expected:   nil,
 		},
 	}
 
 	for _, test := range tests {
-		old, _, err := FindOldReplicaSets(&test.deployment, test.rsList, test.podList)
+		_, old, err := FindOldReplicaSets(&test.deployment, test.rsList)
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util_test.go]: We should test that the RSs are getting filtered correctly. This means that the inputs of this test should contain RSs with both rs.spec.replicas > 0 and rs.spec.replicas == 0 and then test that both outputs here conform to what we expect.

**Comment by shiywang** [comment]
add an new func `TestFindOldReplicaSetsFilter` to test that

**Comment by kargakis** [comment]
@shiywang it doesn't make sense to have two different unit tests for the same unit. Can you collapse the two tests and update the table of the original test with your new expectations?

**Comment by shiywang** [comment]
@kargakis How about now ?

File: pkg/controller/deployment/util/deployment_util.go [hunk 10]
```diff
@@ -796,42 +783,22 @@ func FindNewReplicaSet(deployment *extensions.Deployment, rsList []*extensions.R
 
 // FindOldReplicaSets returns the old replica sets targeted by the given Deployment, with the given PodList and slice of RSes.
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: Comment needs update

**Comment by kargakis** [comment]
One last comment, lgtm otherwise

File: pkg/controller/deployment/util/deployment_util_test.go [hunk 11]
```diff
@@ -297,27 +297,23 @@ func TestGetNewRC(t *testing.T) {
 	}
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util_test.go]: While here can you also rename this test from `TestGetNewRC` to `TestGetNewRS`? Thanks

Comment (shiywang) [inline pkg/controller/deployment/util/deployment_util_test.go]: renamed,  thanks

**Comment by shiywang** [comment]
@k8s-bot cvm gce e2e test this

**Comment by kargakis** [comment]
This shouldn't have merged I think - @shiywang did you look at the test failures? I thought they were related.

**Comment by shiywang** [comment]
@kargakis oh,, sorry, I though it was a flake or sth, so should I revert it back ?

**Comment by kargakis** [comment]
@shiywang you need to make sure it's a known flake that is not affected by your code before you rekick the bot. Those flakes were deployment e2e tests and were flakes most likely caused by this PR. I opened a revert in #43848

**Comment by shiywang** [comment]
@kargakis ok, thanks, I will be cautious of flake and rekick the bot next time, I will run a deployment e2e tests in local and find out the reason of failed

**Pull request merged**